r2dbc-spi/src/main/java/io/r2dbc/spi/ConnectionUrlParser.java

```diff
@@ -0,0 +1,496 @@
+            throw new IllegalArgumentException(String.format("Empty driver in URL: %s", url));
+        }
+    }
+
+    static ConnectionFactoryOptions parseQuery(CharSequence url) {
+
+        String urlToUse = url.toString();
+        validate(urlToUse);
+
+        // R2DBC URL must contain at least two colons in the scheme part (r2dbc:<some driver>:).
+        String[] schemeParts = urlToUse.split(":", 3);
+
+        String scheme = schemeParts[0];
+        String driver = schemeParts[1];
+        String protocol = schemeParts[2];
+
+        int schemeSpecificPartIndex = urlToUse.indexOf("://");
+        String rewrittenUrl = scheme + urlToUse.substring(schemeSpecificPartIndex);
+
+        URI uri = URI.create(rewrittenUrl);
+
+        ConnectionFactoryOptions.Builder builder = ConnectionFactoryOptions.builder();
+
+        if (scheme.equals(R2DBC_SSL_SCHEME)) {
+            builder.option(ConnectionFactoryOptions.SSL, true);
+        }
+
+        builder.option(ConnectionFactoryOptions.DRIVER, driver);
+
+        int protocolEnd = protocol.indexOf("://");
+        if (protocolEnd != -1) {
+            protocol = protocol.substring(0, protocolEnd);
+
+            if (!protocol.trim().isEmpty()) {
+                builder.option(ConnectionFactoryOptions.PROTOCOL, protocol);
+            }
+        }
+
+        if (hasText(uri.getHost())) {
+            builder.option(ConnectionFactoryOptions.HOST, decode(uri.getHost().trim()).toString());
+
+            if (hasText(uri.getRawUserInfo())) {
+                parseUserinfo(uri.getRawUserInfo(), builder);
+            }
+        } else if (hasText(uri.getRawAuthority())) {
+
+            String authorityToUse = uri.getRawAuthority();
+
+            if (authorityToUse.contains("@")) {
+
+                int atIndex = authorityToUse.indexOf('@');
+                String userinfo = authorityToUse.substring(0, atIndex);
+                authorityToUse = authorityToUse.substring(atIndex + 1);
+
+                if (!userinfo.isEmpty()) {
+                    parseUserinfo(userinfo, builder);
+                }
+            }
+
+            builder.option(ConnectionFactoryOptions.HOST, decode(authorityToUse.trim()).toString());
+        }
+
+        if (uri.getPort() != -1) {
+            builder.option(ConnectionFactoryOptions.PORT, uri.getPort());
+        }
+
+        if (hasText(uri.getPath())) {
+            String path = uri.getPath().substring(1).trim();
+            if (hasText(path)) {
+                builder.option(ConnectionFactoryOptions.DATABASE, path);
+            }
+        }
+
+        if (hasText(uri.getRawQuery())) {
+            parseQuery(uri.getRawQuery().trim(), (k, v) -> {
+
+                if (PROHIBITED_QUERY_OPTIONS.contains(k)) {
+                    throw new IllegalArgumentException(
+                        String.format("URL %s must not declare option %s in the query string", url, k));
+                }
+
+                builder.option(Option.valueOf(k), v);
```

Comment:
When query has multiple values with same key, this will override the value to the last one.
e.g.: `?foo=aaa&foo=bbb&foo=ccc` -> Option("foo","ccc")

Need to handle multiple values as set?

Reply:
Right now, multiple values are overwritten by the last value. Let's keep that option open and do this at a later time. 

I will add a note to the spec that consumers of an option should expect a potentially different type than `T` as in `Option<T>`.